src/AMAMM.sol

```diff
@@ -2,69 +2,244 @@
+    mapping(PoolId id => mapping(uint40 => mapping(address => Bid))) public poolEpochBids;
+
+    /// -----------------------------------------------------------------------
+    /// Getter actions
+    /// -----------------------------------------------------------------------
+
+    function getManager(PoolId id, uint40 epoch) public view returns (Bid memory) {
+        return poolEpochManager[id][epoch];
+    }
+
+    function getBid(PoolId id, uint40 epoch, address bidder) public view returns (Bid memory) {
+        return poolEpochBids[id][epoch][bidder];
+    }
+
+    /// -----------------------------------------------------------------------
+    /// Bidder actions
+    /// -----------------------------------------------------------------------
+    /// @inheritdoc IAmAmm
+    function bid(PoolId id, address bidder, bytes7 payload, uint128 rent, uint128 deposit, uint40 _epoch)
+        external
+        virtual
         override
-        returns (Hooks.Permissions memory)
     {
-        return
-            Hooks.Permissions({
-                beforeInitialize: false,
-                afterInitialize: false,
-                beforeAddLiquidity: false,
-                afterAddLiquidity: false,
-                beforeRemoveLiquidity: false,
-                afterRemoveLiquidity: false,
-                beforeSwap: false,
-                afterSwap: true, // Override how swaps are done
-                beforeDonate: false,
-                afterDonate: false,
-                beforeSwapReturnDelta: false,
-                afterSwapReturnDelta: true, // Allow afterSwap to return a custom delta
-                afterAddLiquidityReturnDelta: false,
-                afterRemoveLiquidityReturnDelta: false
-            });
+        /// -----------------------------------------------------------------------
+        /// Validation
+        /// -----------------------------------------------------------------------
+
+        address msgSender = LibMulticaller.senderOrSigner();
+
+        // ensure bid is valid
+        // - manager can't be zero address
+        // - bid needs to be greater than the next bid by >10%
+        // - deposit needs to cover the rent for K hours
+        // - deposit needs to be a multiple of rent
+        // - payload needs to be valid
+        if (
+            bidder == address(0) || rent <= poolEpochManager[id][_epoch].rent.mulWad(MIN_BID_MULTIPLIER(id))
+                || deposit < rent * K(id) || deposit % rent != 0 || !_payloadIsValid(id, payload)
+        ) {
+            revert AmAmm__InvalidBid();
+        }
+
+        // Check if the bid already exists
+        Bid memory existingBid = poolEpochBids[id][_epoch][bidder];
+        Bid memory newBid = Bid({bidder: msgSender, payload: payload, rent: rent, deposit: deposit});
+
+        if (existingBid.bidder != address(0)) {
+            // If the bid exists, update the deposit
+            existingBid.deposit += deposit;
```

Comment:
Here, we should process the rent deposits. It should reduce the deposit balance for the new high bidder and increase the balance for the outbidder to treat as a refund.